lfs/gitscanner_tree.go

```diff
@@ -41,38 +41,20 @@ func catFileBatchTree(treeblobs *TreeBlobChannelWrapper) (*PointerChannelWrapper
 	errchan := make(chan error, 10) // Multiple errors possible
 
 	go func() {
+		scanner := &catFileBatchScanner{r: cmd.Stdout}
 		for t := range treeblobs.Results {
 			cmd.Stdin.Write([]byte(t.Sha1 + "\n"))
-			l, err := cmd.Stdout.ReadBytes('\n')
-			if err != nil {
-				break
-			}
-
-			// Line is formatted:
-			// <sha1> <type> <size>
-			fields := bytes.Fields(l)
-			s, _ := strconv.Atoi(string(fields[2]))
-
-			nbuf := make([]byte, s)
-			_, err = io.ReadFull(cmd.Stdout, nbuf)
-			if err != nil {
-				break // Legit errors
-			}
 
-			p, err := DecodePointer(bytes.NewBuffer(nbuf))
-			if err == nil {
-				pointers <- &WrappedPointer{
-					Sha1:    string(fields[0]),
-					Pointer: p,
-					Name:    t.Filename,
-				}
-			}
-
-			_, err = cmd.Stdout.ReadBytes('\n') // Extra \n inserted by cat-file
+			p, err := scanner.Next()
 			if err != nil {
+				errchan <- err
 				break
+			} else if p != nil {
+				p.Name = t.Filename
+				pointers <- p
```

Comment:
This is an interesting pattern. This command synchronously writes to stdin, and then reads the response from stdout. This is so it can associate the returned pointer with the given filename from `ls-tree`. This is incompatible with `Scan()`, since it will keep reading until there's a _valid_ lfs pointer.

I think this synchronous approach is a solid pattern though. Pretty much every scan call needs to associate the filename at least. Other funcs do this by keeping some map of shas and filenames, which is :-1:.